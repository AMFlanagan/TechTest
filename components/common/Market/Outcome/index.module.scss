.outcome {
  // reset button styles
  background: none;
  color: inherit;
  border: none;
  padding: 0;
  font: inherit;
  cursor: pointer;
  outline: inherit;

  // styles
  display: flex;
  flex-direction: column;
  justify-content: space-evenly;
  background-color: #222222;
  text-align: center;
  width: 100%;
  padding: 16px;
}

.selected,
.outcome:hover {
  background-color: green;
}

.odds,
.name {
  padding: 4px 0;
  align-self: center;
}
