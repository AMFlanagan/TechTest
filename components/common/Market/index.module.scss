.marketContainer {
  background-color: #555555;
}

.marketName {
  text-align: center;
  padding: 4px;
}

.outcomes {
  display: flex;
  flex-direction: column;
  justify-content: space-evenly;
}

.outcomes > * {
  border-bottom: 2px solid #555555;
}

.outcomes > *:last-child {
  border: none;
}

@media only screen and (min-width: 600px) {
  .outcomes {
    display: flex;
    flex-direction: row;
    justify-content: space-evenly;
  }

  .outcomes > * {
    border: none;
    border-right: 2px solid #555555;
  }

  .outcomes > *:last-child {
    border: none;
  }
}


