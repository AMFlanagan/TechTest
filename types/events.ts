export enum Sport {
  FOOTBALL = 'football',
  RUGBY = 'rugby',
  HORSE_RACING = 'horse-racing',
}

export enum FootballMarket {
  RESULT = 'Result',
  NEXT_GOAL = 'Next goal',
}

export enum RugbyMarket {
  RESULT = 'Result',
  NEXT_TRY = 'Next Try',
}

export enum HorseRacingMarket {
  RESULT = 'Result',
}

export interface IOutcome {
  id: number
  name: string
  odds: number
}

export type Market = FootballMarket | HorseRacingMarket | RugbyMarket

export interface IMarket {
  type: Market
  outcomes: IOutcome[]
}

export interface IEvent {
  id: number
  sport: Sport
  eventName: string
  markets: IMarket[]
  startTime: string
}
