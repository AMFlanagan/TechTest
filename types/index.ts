export interface IEvent {
  marketId: number
  outcomes: string[]
}
