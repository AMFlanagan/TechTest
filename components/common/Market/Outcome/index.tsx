import React from 'react'
import { IOutcome } from '../../../../types'

import formatNumbers from '../../../../utils/formatNumbers'
import { useBetSlip } from '../../../../context/BetSlip'

import styles from './index.module.scss'

interface IProps {
  outcome: IOutcome
  onSelect: (
    outcomeId: number,
    outcomeName: string,
    outcomeOdds: number,
    inBetSlip: boolean
  ) => void
}

const Outcome: React.FC<IProps> = (props: IProps) => {
  const { state } = useBetSlip()

  const isInBetSlip = () =>
    !!state.bets.find((bet: any) => bet.outcomeId === props.outcome.id)

  return (
    <button
      onClick={() =>
        props.onSelect(
          props.outcome.id,
          props.outcome.name,
          props.outcome.odds,
          isInBetSlip()
        )
      }
      className={`${styles.outcome} ${isInBetSlip() && styles.selected}`}
    >
      <div className={styles.name}>{props.outcome.name}</div>
      <div className={styles.odds}>{formatNumbers(props.outcome.odds)}</div>
    </button>
  )
}

Outcome.displayName = 'Outcome'

export default Outcome
