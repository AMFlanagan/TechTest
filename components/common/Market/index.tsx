import React from 'react'
import { IMarket } from '../../../types'
import Outcome from '../Outcome'

import styles from './index.module.scss'

interface IProps {
  market: IMarket
}

const Market: React.FC<IProps> = (props: IProps) => {
  return (
    <div className={styles.marketContainer}>
      <div className={styles.marketName}>{props.market.type}</div>
      <div className={styles.outcomes}>
        {props.market.outcomes.slice(0, 4).map((outcome, i: number) => {
          return <Outcome key={i} outcome={outcome} />
        })}
      </div>
    </div>
  )
}

Market.displayName = 'Market'

export default Market
