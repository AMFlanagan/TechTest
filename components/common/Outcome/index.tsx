import React from 'react'
import { IOutcome } from '../../../types'
import formatOdds from '../../../utils/formatOdds'

import styles from './index.module.scss'

interface IProps {
  outcome: IOutcome
}

const Outcome: React.FC<IProps> = (props: IProps) => {
  return (
    <button className={styles.outcome}>
      <div className={styles.name}>{props.outcome.name}</div>
      <div className={styles.odds}>{formatOdds(props.outcome.odds)}</div>
    </button>
  )
}

Outcome.displayName = 'Outcome'

export default Outcome
