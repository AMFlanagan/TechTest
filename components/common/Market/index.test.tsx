import { render } from '@testing-library/react'

import Market from '.'
import { FootballMarket, IMarket } from '../../../types'
import Outcome from './Outcome'

jest.mock('./Outcome', () => jest.fn(() => <div></div>))

const dispatchMock = jest.fn()

const mockHook = {
  dispatch: dispatchMock,
}

jest.mock('../../../context/BetSlip', () => ({
  useBetSlip() {
    return mockHook
  },
}))

describe('Market', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })

  it('renders as expect', () => {
    const market: IMarket = {
      id: 1,
      outcomes: [
        {
          id: 1,
          name: 'home',
          odds: 1.0,
        },
        {
          id: 2,
          name: 'away',
          odds: 2.0,
        },
      ],
      type: FootballMarket.RESULT,
    }

    const { asFragment, getByText } = render(
      <Market market={market} eventName={''} eventId={0} />
    )

    expect(getByText(FootballMarket.RESULT)).toBeTruthy()

    market.outcomes.forEach((outcome) => {
      expect(Outcome).toBeCalledWith(
        expect.objectContaining({
          outcome,
        }),
        expect.anything()
      )
    })

    expect(asFragment()).toMatchSnapshot()
  })

  it('calls Outcome a maximum of 4 times', () => {
    const market: IMarket = {
      id: 1,
      outcomes: [
        {
          id: 1,
          name: 'home',
          odds: 1.0,
        },
        {
          id: 2,
          name: 'away',
          odds: 2.0,
        },
        {
          id: 3,
          name: 'draw',
          odds: 1.0,
        },
        {
          id: 4,
          name: 'none',
          odds: 2.0,
        },
        {
          id: 5,
          name: 'none',
          odds: 2.0,
        },
      ],
      type: FootballMarket.RESULT,
    }

    render(<Market market={market} eventName={''} eventId={0} />)
    expect(Outcome).toHaveBeenCalledTimes(4)
  })
})
