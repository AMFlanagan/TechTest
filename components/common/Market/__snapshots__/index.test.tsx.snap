// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Market renders as expect 1`] = `
<DocumentFragment>
  <div
    class="marketContainer"
  >
    <div
      class="marketName"
    >
      Result
    </div>
    <div
      class="outcomes"
    >
      <div />
      <div />
    </div>
  </div>
</DocumentFragment>
`;
